packages/world/src/interfaces/systems/IRegistrationSystem.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { IStoreHook } from "@latticexyz/store/src/Store.sol";
+import { Schema } from "@latticexyz/store/src/Schema.sol";
+
+import { System } from "../../System.sol";
+
+import { ISystemHook } from "../ISystemHook.sol";
+
+// TODO: autogenerate
+
+interface IRegistrationSystem {
+  /**
+   * Register a new namespace
+   */
+  function registerNamespace(bytes16 namespace) external;
+
+  /**
+   * Register a table with given schema in the given namespace
+   */
+  function registerTable(
+    bytes16 namespace,
+    bytes16 file,
+    Schema keySchema,
+    Schema valueSchema
+  ) external returns (bytes32 resourceSelector);
+
+  /**
+   * Register metadata (tableName, fieldNames) for the table at the given namespace and file.
+   * Requires the caller to own the namespace.
+   */
+  function setMetadata(
+    bytes16 namespace,
+    bytes16 file,
+    string calldata tableName,
+    string[] calldata fieldNames
+  ) external;
+
+  /**
+   * Register the given store hook for the table at the given namespace and file.
+   * Hooks on table files must implement the IStoreHook interface,
+   * and hooks on system files must implement the ISystemHook interface.
+   */
+  function registerHook(bytes16 namespace, bytes16 file, address hook) external;
+
+  /**
+   * Register a hook for the table at the given namepace and file.
+   * Requires the caller to own the namespace.
+   */
+  function registerTableHook(bytes16 namespace, bytes16 file, IStoreHook hook) external;
+
+  /**
+   * Register a hook for the system at the given namespace and file
+   */
+  function registerSystemHook(bytes16 namespace, bytes16 file, ISystemHook hook) external;
+
+  /**
+   * Register the given system in the given namespace.
+   * If the namespace doesn't exist yet, it is registered.
+   * The system is granted access to its namespace, so it can write to any table in the same namespace.
+   * If publicAccess is true, no access control check is performed for calling the system.
+   */
+  function registerSystem(
+    bytes16 namespace,
+    bytes16 file,
+    System system,
+    bool publicAccess
+  ) external returns (bytes32 resourceSelector);
+
+  /**
+   * Register a World function selector for the given namespace, file and system function.
+   */
+  function registerFunctionSelector(
+    bytes16 namespace,
+    bytes16 file,
+    string memory systenFunctionName,
+    string memory systemFunctionArguments
+  ) external returns (bytes4 worldFunctionSelector);
+
+  /**
+   * Register a root World function selector (without namespace / file prefix).
+   */
+  function registerRootFunctionSelector(
+    bytes16 namespace,
+    bytes16 file,
+    bytes4 worldFunctionSelector,
+    bytes4 systemFunctionSelector
+  ) external returns (bytes4 functionSelector);
```

Comment:
Does access control use the full World method path (`{namespace}_{file}_{functionName}`) as part of its access control? or does it just look up the function by name to get the registration info (namespace, file) and does access control based on that?

If the latter, I wonder if we could allow for _any_ kind of function name registered on the world (maybe excluding a `_` prefix for internals), not just a namespaced one, mostly to make it a little more user friendly? We're already enforcing uniqueness.

(I realize we could autogen around this but maybe we don't need to)

Reply:
it's the latter, so we could certainly allow for other formats of function selectors to be registered, like we do in `registerRootFunctionSelector`, but I'd prefer to leave those fully flexible function selectors exclusive to the owner of the root namespace to avoid "squatting" on important function selectors. We could allow a custom function prefixed by only the namespace though!

Reply:
However I kind of like that with the verbose function selector it's immediately visible where the selector maps to (namespace, file and function selector within the system), while with custom ones it might be harder to know where to look for the source code